### PR TITLE
KAFKA:5653: add join overloads to KTable

### PR DESCRIPTION
**Pull request by dguy**
Add `join`, `leftJoin`, `outerJoin` overloads that use `Materialized` to `KTable`

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang, @bbejeck, @mjsax

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 1]
```diff
@@ -1342,6 +1342,81 @@ void to(final Serde<K> keySerde,
      * Both input streams (or to be more precise, their underlying source topics) need to have the same number of
      * partitions.
      *
+     * @param other         the other {@code KTable} to be joined with this {@code KTable}
+     * @param joiner        a {@link ValueJoiner} that computes the join result for a pair of matching records
+     * @param materialized  an instance of {@link Materialized} used to describe how the state store should be materialized
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: Nit: add `Cannot be {@code null}.` -- Guess also somewhere else

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 2]
```diff
@@ -1603,7 +1765,9 @@ void to(final Serde<K> keySerde,
      * left {@code KTable}
      * @see #join(KTable, ValueJoiner)
      * @see #outerJoin(KTable, ValueJoiner)
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: Update both references ?

**Review by mjsax — APPROVED** [review]
Just some nits. Overall LGTM.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 3]
```diff
@@ -556,7 +566,14 @@ public V apply(Change<V> change) {
...
                                           final ValueJoiner<? super V, ? super V1, ? extends R> joiner) {
-        return doJoin(other, joiner, true, true, null, null);
+        return doJoin(other, joiner, null, true, true);
+    }
+
+    @Override
+    public <VO, VR> KTable<K, VR> outerJoin(final KTable<K, VO> other,
+                                            final ValueJoiner<? super V, ? super VO, ? extends VR> joiner,
+                                            final Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        return doJoin(other, joiner, new MaterializedInternal<>(materialized), true, true);
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:576]: nit: single parameter per line

**Comment by bbejeck** [comment]
one other nit: not from this PR but lines 87, 109, 147 and 187 have `dot not satisfied` -> `do not satisfy` 

Otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:576]: I think it is fine, as the whole call statement itself is not long (we set it to sth. like 120 char?)

**Review by guozhangwang — COMMENTED** [review]
LGTM.

**Comment by dguy** [comment]
Merged to trunk

**Pull request closed without merging**